src/rust/engine/fs/src/glob_matching.rs

```diff
@@ -157,246 +125,171 @@ trait GlobMatchingImplementation<E: Display + Send + Sync + 'static>: VFS<E> {
       return future::ok(vec![]).to_boxed();
     }
 
-    let init = PathGlobsExpansion {
-      context: self.clone(),
-      todo: include
-        .iter()
-        .flat_map(|entry| entry.to_sourced_globs())
-        .collect(),
-      exclude,
-      completed: IndexMap::default(),
-      outputs: IndexSet::default(),
-    };
-    future::loop_fn(init, |mut expansion| {
-      // Request the expansion of all outstanding PathGlobs as a batch.
-      let round = future::join_all({
-        let exclude = &expansion.exclude;
-        let context = &expansion.context;
-        expansion
-          .todo
-          .drain(..)
-          .map(|sourced_glob| context.expand_single(sourced_glob, exclude))
-          .collect::<Vec<_>>()
-      });
-      round.map(move |single_expansion_results| {
-        // Collect distinct new PathStats and PathGlobs
-        for exp in single_expansion_results {
-          let SingleExpansionResult {
-            sourced_glob: GlobWithSource { path_glob, source },
-            path_stats,
-            globs,
-          } = exp;
+    let result = Arc::new(Mutex::new(IndexSet::default()));
```

Comment:
It seems strange having an `IndexSet` inside a `Mutex`; the `Mutex` suggests it will be written to from multiple threads, but the `IndexSet` suggests we care about ordering of writes.

What ordering are we aiming for? It feels like we should probably have each future return a `Vec`, and then merge them into an `IndexSet` one at a time root by root?

Reply:
Good idea!

So: while it would be very functional to recursively merge vecs as output from these functions, it felt like that would likely result in a bunch of allocation that could potentially be more expensive than the lock acquisition. I expect that we'll do further refactoring in here now that it is simplified, so I left this using the `Mutex` accumulator, but switched it to collecting into a Vec, and sorting/deduping at the end (since I expect that dupes will be rare).

Reply:
Sounds good :)

The other option here would be to accumulate in a `Arc<Mutex<BTreeSet>>` and then just `result.into_iter().collect()` - I have no real preference :)